### ISSUE TITLE
More of a Question than an Issue

### ISSUE THREAD

**Issue by ilay32** [Wolframm74/armadillo_armanpy#2]
Hi, I have been fiddling about with armanpy directly. It builds ok for python2.7. 
But since this is supposed to fit into a pretty large phython3 program i'm working on, I tried to build with flags pointing to my python3.6 libraries. The cmake --build command then fails with numerous errors. My question is:
1. Can you say if and how it would be possible to build the given example for python3 ?
2. Will using your installer solve this issue and allow me to integrate some armadillo in my python3 program?
I'm asking you and not the armanpy authors, since it seems they haven't touched this code for some years while your work is pretty recent.
Thanks,
Ilay

**Comment by Wolframm74** [comment]
I can try to re-compile for python3 but I need a bit of time. I haven't been active on here in a while either. I'll try to reply within a few hours.

**Comment by ilay32** [comment]
Cheers, no worries and no hurries either. In this particular case i just ended up using lapacke and cblas and extending python directly. But an armadillo python 3  wrapper could sure be useful in other cases. If you can hack it you'll be my hero.

**Comment by Wolframm74** [comment]
Very nice!!! @ilay32 Please try the build again with

cmake -DPYTHON3=1 .

from the top level of armadillo_armanpy. It works for me on my system and appears to be using python3. If it works in yours let me know and if not please post the build output or the issue you are facing.

**Comment by Wolframm74** [comment]
Just curious, do you absolutely need to use SWIG - which is somewhat complicated to maintain or would you be moreso interested in extending something like the LLS example library? I personally think the latter would be easier, but some of the pieces of the SWIG and armanpy code are there in case you need to consider either route. SWIG is not something I use day-to-day so I am generally less familiar with it.

**Comment by ilay32** [comment]
OK, it builds, but seems to still target python2 . And indeed the tests and example in armadillo_armanpy/armanpy-0.1.4/build/bin run only with python2.7 (for some reason the tests don't pass, but the script runs).

this is the build output:
[make.log](https://github.com/Wolframm74/armadillo_armanpy/files/4942762/make.log)

there are also some warnings (yes SWIG...) :
[makerr.log](https://github.com/Wolframm74/armadillo_armanpy/files/4942763/makerr.log)

About SWIG -- I never used it explicitly. It is just part of the armanpy way. I can make a python extension manually if need to, and there I can use whatever I want. But my thought at the time was to use armadillo and let armanpy stitch it to python.
Personally for me it's not very important at the moment at all. It is quite rarely that I find it necessary to boost performance this way. I will however try to look into it when I get the time. maybe i'll learn something, and with a bit of luck i might even be able to make a useful contribution :)

**Comment by Wolframm74** [comment]
Thanks!!! ... Based on  your response I have added the use of the -DPYTHON3=1 build parameter into the build of armanpy, which was missing before. You can see this in f716cc62f0ba7fd06976cf1f1977d89af268a371. Now, there are a bunch of c++ errors inherent in the build of the library that will have to be looked at. Upon first glance they appear pretty swiggy.... I also noted that yes some of the tests in python 2.7 don't pass as you pointed out.

**Comment by ilay32** [comment]
yes I suppose these are the same issues that popped when I tried to build armanpy for python3 without this installer (see my original question). The authors have not touched it in some years. I guess a revision of both the swig usage (see deprecation warnings) and the python extension mechanism (through the swig templates) is required.  probably not worth the trouble, but i'll take a crack at it too if i feel cracky

**Comment by Wolframm74** [comment]
I may have time to look into it, but again no guarantees and that is fully correct, I am not the originator of the armanpy code. 

Meanwhile, have a look at this branch code (da5699ec1293f1bf4bd7470af8884089bde3af87) if you are searching for ideas on how to hook up this library to python.

https://github.com/Wolframm74/armadillo_armanpy/blob/feat/python-interface/LLS/build/LLS_lib/lls_test.py
The work on that branch (python-interface) is inspired by the following blog post:
https://nesi.github.io/perf-training/python-scatter/ctypes

It may be easier than dealing with swig templates.

**Comment by ilay32** [comment]
Wow this looks much simpler. I'll bear in mind next time i need to integrate C. I guess sorting out armanpy for python3 remains as a sporty challenge only. There is really no practical need for it anymore is there.

**Comment by Wolframm74** [comment]
Yes, it is likely so (simpler). By chance I had that branch in a local repository in a different PC at home, while working on a different path. Depending what you need for your project, using python would be a matter of coming to a stable concept of what functions your library exposes and then doing something like that to make it visible to python2 or python3. The only thing you'll need to agree upon is matrix sizes. With armadillo each time you refactor the code you have to tell the compiler an idea of how big the matrix sizes it will be dealing with are. 

For instance, here

```
mylib.lls_solver(x_array, y_array, 4, b_array, 2)
```

It is pre-determined that I am solving a matrix equation with problem size 4 - the LLS equation has vectors of size N=4. So you will see that all through-out my code. I think it can be made flexible depending on what exactly you need to do for your matrix problem.
